include/core/CBlockingCallCancellingTimer.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+#ifndef INCLUDED_ml_core_CBlockingCallCancellingTimer_h
+#define INCLUDED_ml_core_CBlockingCallCancellingTimer_h
+
+#include <core/CBlockingCallCancellerThread.h>
+#include <core/CoreTypes.h>
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+
+namespace ml {
+namespace core {
+
+//! \brief
+//! Cancels a blocking IO call after a supplied timeout has elapsed.
+//!
+//! DESCRIPTION:\n
+//! A thread that, when started, waits for the specified number of
+//! seconds, then cancels any in-progress blocking IO call in the
+//! other thread specified on construction.
+//!
+//! If this thread is shut down before the timeout occurs then no
+//! blocking IO call cancellation takes place.
+//!
+//! IMPLEMENTATION DECISIONS:\n
+//! Designed to handle the following sequence of events:
+//! 1) Java process spawns this process
+//! 2) This process creates named pipes and waits for Java to connect
+//! 3) Java never connects for some reason
+//!
+//! We want to unblock this process when this happens so that it can
+//! delete the named pipe and exit.
+//!
+class CBlockingCallCancellingTimer : public CBlockingCallCancellerThread {
+public:
+    //! Default timeout if not specified.
+    static const ml::core_t::TTime DEFAULT_TIMEOUT_SECONDS;
```

Comment:
I decided to stick with `TTime` here rather than switch to `std::chrono::duration`.  The reason is that all the existing command line argument parsing is currently using `TTime` and taking whole numbers of seconds, and it would be inconsistent if the new command line argument tried to do something clever with generic durations.

I think at some point we do need to move over to the `std::chrono` stuff throughout the codebase.  This is obviously a huge project but unless we do it we're going to have messy boundary points like this.  It would also allow us to support sub-second bucket spans more naturally than the scaling approach previously envisaged.  However, the time to do it is probably once we've upgraded to C++20 when the standard library will have support for parsing durations from text (as well as support for parsing timestamps - see #1459).